include/operations/blas3_trees_gemm.hpp

```diff
@@ -0,0 +1,674 @@
+   *                 for details
+   * @param in_row  a predicate which checks whether a row index is within
+   *                matrix bounds
+   * @param in_col  a predicate which checks whether a col index is within
+   *                matrix bounds
+   */
+  template <bool check_row_limit, bool check_col_limit, bool trans,
+            int rows, int cols, int lds,
+            typename InputPointerType, typename ScratchPointerType,
+            typename RowPredicate, typename ColPredicate>
+  static inline typename std::enable_if<!trans>::type
+  extract_block(
+      int item_id, InputPointerType ptr, int ld, ScratchPointerType scratch,
+      RowPredicate in_row, ColPredicate in_col)
+  {
+    const int bs = rows * cols;
+    #pragma unroll
+    for (int i = 0; i < (bs - 1) / wg_size + 1; ++i) {
+      if (!do_check<bs % wg_size>(item_id + i*wg_size < bs)) continue;
+      const int col_ofs = i * (wg_size/rows);
+      const bool in_range =
+          do_check<check_row_limit>(in_row(item_id%rows, 0)) &&
+          do_check<check_col_limit>(in_col(item_id/rows, col_ofs));
+      scratch[col_ofs * lds] = in_range ? ptr[col_ofs * ld] : T(0);
+    }
+  }
+
+  template <bool check_row_limit, bool check_col_limit, bool trans,
+            int rows, int cols, int lds,
+            typename InputPointerType, typename ScratchPointerType,
+            typename RowPredicate, typename ColPredicate>
+  static inline typename std::enable_if<trans>::type
+  extract_block(
+      int item_id, InputPointerType ptr, int ld, ScratchPointerType scratch,
+      RowPredicate in_row, ColPredicate in_col)
+  {
+    const int bs = rows * cols;
+    #pragma unroll
+    for (int i = 0; i < (bs - 1) / wg_size + 1; ++i) {
+      if (!do_check<bs % wg_size>(item_id + i*wg_size < bs)) continue;
+      const int row_ofs = i * (wg_size/cols);
+      const bool in_range =
+          do_check<check_row_limit>(in_row(item_id/cols, row_ofs)) &&
+          do_check<check_col_limit>(in_col(item_id%cols, 0));
+      scratch[row_ofs] = in_range ? ptr[row_ofs * ld] : T(0);
+    }
+  }
+
+  /*!
+   * @brief Compute a small matrix-matrix product `reg_res += A*B`.
+   *
+   * @tparam InputPointerType  pointer type for A and B
+   *
+   * @param B  pointer to matrix A with proper item-dependent offset,
+   *           see GemmFactory::run() for details
+   * @param A  pointer to matrix B with proper item-dependent offset,
+   *           see GemmFactory::run() for details
+   * @param reg_a  temporary register array used to prefetch columns of A
+   * @param reg_b  temporary register used to prefetch elements of B
+   * @param reg_res  2D register array used to store the result C
+   */
+  template <typename InputPointerType>
+  static inline void compute_block_gemm(
+      InputPointerType B, InputPointerType A,
+      T (&reg_a)[item_rows], T &reg_b, T (&reg_res)[item_rows][item_cols])
+      noexcept
+  {
+    // NOTE: adding pragma unroll here reduces performance
```

Comment:
Can you give maybe some more information? e.g was it there a particular reason or was it on a particular platform?